File: Quick-Sort/js/leetcode_lc_912.js
const nums = [5,2,3,1]

var sortArray = function(nums) {
    
    quicksort(nums,0,nums.length-1)

    return nums
};


function quicksort(arr,start,end){
    if(start >= end){
        return 
    }

    const pivotIdx = partition(arr,start,end)

    quicksort(arr,start,pivotIdx-1)
    quicksort(arr,pivotIdx+1,end)
}

function partition(arr,start,end){
    let leftIdx = start , rightIdx = end
    const pivot = arr[start]

    while(leftIdx < rightIdx){
        while(arr[rightIdx] > pivot && leftIdx < rightIdx){
            rightIdx--
        }

        while(arr[leftIdx] <= pivot && leftIdx < rightIdx){
            leftIdx++
        }

        if(leftIdx < rightIdx){
            [arr[leftIdx],arr[rightIdx]] = [arr[rightIdx],arr[leftIdx]]
        }
    }

    arr[start] = arr[leftIdx]
    arr[leftIdx] = pivot
    return leftIdx
}



console.log(sortArray(nums))

File: Quick-Sort/python/leetcode_lc_912.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) <= 1:
            return nums

        # 随机取数 避免因为pivot区分度不强造成的算法退化
        pivot = random.choice(nums)

        # O(n)划分
        left  = self.sortArray([x for x in nums if x < pivot])
        right = self.sortArray([x for x in nums if x > pivot])
        
        # 相同值保留 避免因为大量相同元素造成的算法退化
        mid  = [x for x in nums if x == pivot]

        return left + mid + right 
        